### PR TITLE
Expose `AuthorizationError`

### PR DESCRIPTION
**Pull request by khieta**
## Description of changes

* Publicly exposed the `EvaluationError` and `AuthorizationError` error types.
* Changed the `errors` field of the `Diagnostics` type from `HashMap<String>` to `Vec<AuthorizationError>`. 
  * A `HashMap` would probably be more accurate, since users shouldn't depend on the ordering of errors, but this requires `AuthorizationError` to be hashable, which is a pain since the errors can contains `Expr`s, `Value`s, other `HashMap`s, etc.
* Removed some unused code in `integration_testing.rs`. The code was checking for a parse error where only an AuthorizationError can be returned.
* Changed `Response` to not be (de)serializable. The only code that relies on this being the case is in `src/frontend`. Defined a new `InterfaceResponse` type there to make things simpler.
  * It would also be possible to make `Response` (de)serializable directly, but this is a pain for the same reason given above for hashing.

This is a major breaking change since it changes the public APIs `Diagnostics::errors()` and `Response::new()`, the public type `EvaluationError`, and the traits derived for `Response`.

I expect a downstream failure in `cedar-spec/cedar-drt` since `Response`  is no longer (de)serializable. I will have a PR up with a fix before I merge this PR.

## Issue #, if available

#114

## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [X] A breaking change requiring a major version bump to `cedar-policy` (e.g., changes to the signature of an existing API).

I confirm that this PR (choose one, and delete the other options):

- [X] Updates the "Unreleased" section of the CHANGELOG with a description of my change (required for major/minor version bumps).

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- [X] Requires updates, and I have made / will make these updates myself. (Please include in your description a timeline or link to the relevant PR in `cedar-spec`, and how you have tested that your updates are correct.)

## Disclaimer

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by khieta** [comment]
Marking as review until I have time to resolve the issue with the test case on L602 of `frontend/is_authorized.rs`. Happy to take any feedback in the meantime. Especially interested in a review from @andrewmwells-amazon to check that I'm not breaking the FFI/Java with my changes.

File: cedar-policy/src/frontend/is_authorized.rs [hunk 1]
```diff
@@ -646,8 +695,16 @@ mod test {
...
-                assert_eq!(response.decision(), Decision::Deny);
+                let parsed_result: AuthorizationAnswer =
+                    serde_json::from_str(result.as_str()).unwrap();
+                match parsed_result {
+                    AuthorizationAnswer::ParseFailed { .. } => {
+                        panic!("expected parse to succeed, but got {parsed_result:?}")
+                    }
+                    AuthorizationAnswer::Success { response } => {
+                        assert_eq!(response.decision, Decision::Deny);
+                    }
```
Comment (cdisselkoen) [inline cedar-policy/src/frontend/is_authorized.rs:706]: at line 685 we also assert there were no errors, should we do the same here?

**Comment by khieta** [comment]
PRs up to fix `cedar-spec` and `cedar-java`: [cedar-spec#101](https://github.com/cedar-policy/cedar-spec/pull/101), [cedar-java#34](https://github.com/cedar-policy/cedar-java/pull/34). Both PRs (and this one) are currently marked as drafts. Hopefully all three will be ready for review later today.

**Comment by khieta** [comment]
Marking this PR and the `cedar-spec` one as "ready for review". The `cedar-java` PR is still in progress.

**Review by cdisselkoen — APPROVED** [review]

File: cedar-policy/src/frontend/is_authorized.rs [hunk 2]
```diff
@@ -51,25 +55,85 @@ fn is_authorized(call: AuthorizationCall) -> Response {
...
+
+impl InterfaceDiagnostics {
+    /// Get the policies that contributed to the decision
+    pub fn reason(&self) -> impl Iterator<Item = &PolicyId> {
+        self.reason.iter()
+    }
+
+    /// Get the errors
+    pub fn errors(&self) -> impl Iterator<Item = String> + '_ {
+        self.errors.iter().cloned()
```
[the remark this answers is not in the document]
Comment (khieta) [reply cedar-policy/src/frontend/is_authorized.rs]: Fixed in the latest commit

**Review by aaronjeline — APPROVED** [review]

**Pull request merged**